### PR TITLE
esp_http_client: fix memory leak and allow set authtype (IDFGH-2315)

### PR DESCRIPTION
**Pull request by AxelLin**
Address below issue:
https://github.com/espressif/esp-idf/issues/4444

### CODE REVIEW

**Comment by Alvin1Zhang** [comment]
@AxelLin Thanks for the contribution.

File: components/esp_http_client/esp_http_client.c [hunk 1]
```diff
@@ -325,16 +327,24 @@ esp_err_t esp_http_client_set_password(esp_http_client_handle_t client, char *pa
...
         client->connection_info.password = NULL;
     } else if (password != NULL) {
         client->connection_info.password = strdup(password);
     }
     return ESP_OK;
 }
 
+esp_err_t esp_http_client_set_authtype(esp_http_client_handle_t client, esp_http_client_auth_type_t auth_type)
+{
+    client->connection_info.auth_type = auth_type;
```
Comment (mahavirj) [inline components/esp_http_client/esp_http_client.c]: Recommend adding check for validity of `client` handle.

File: components/esp_http_client/esp_http_client.c [hunk 2]
```diff
@@ -300,8 +300,10 @@ esp_err_t esp_http_client_set_username(esp_http_client_handle_t client, const ch
         ESP_LOGE(TAG, "client must not be NULL");
         return ESP_ERR_INVALID_ARG;
     }
-    if (username == NULL && client->connection_info.username != NULL) {
+    if (client->connection_info.username != NULL) {
         free(client->connection_info.username);
+    }
+    if (username == NULL) {
```
Comment (mahavirj) [inline components/esp_http_client/esp_http_client.c]:
```suggestion
    client->connection_info.username = username ? strdup(username) : NULL;
```

File: components/esp_http_client/esp_http_client.c [hunk 3]
```diff
@@ -325,16 +327,24 @@ esp_err_t esp_http_client_set_password(esp_http_client_handle_t client, char *pa
         ESP_LOGE(TAG, "client must not be NULL");
         return ESP_ERR_INVALID_ARG;
     }
-    if (password == NULL && client->connection_info.password != NULL) {
+    if (client->connection_info.password != NULL) {
         memset(client->connection_info.password, 0, strlen(client->connection_info.password));
         free(client->connection_info.password);
+    }
+    if (password == NULL) {
```
Comment (mahavirj) [inline components/esp_http_client/esp_http_client.c]:
```suggestion
    client->connection_info.password = password ? strdup(password) : NULL;
```

**Pull request closed without merging**